Unable to deserialize a string value class properly

**Describe the bug**
The confluent's deserializer, when the writer schema is a STRING, it always applies `result = result.toString()` which finally shows a ClassCastException as the app code receives the wrong type.

**Stack trace**
```
Caused by: java.lang.ClassCastException: class java.lang.String cannot be cast to class MyType
```

**To Reproduce**
Code to reproduce the behavior:
```kotlin
@Serializable
@JvmInline
value class MyWrapper(val value: String)

val schema = Schema.create(Schema.Type.STRING)
val schemaRegistry = MockSchemaRegistryClient()
val bytes = KafkaAvroSerializer(schemaRegistry).serialize("topic", "the value")
val decodedValue = SpecificAvro4kKafkaDeserializer(isKey = true, schemaRegistry = schemaRegistry).deserialize("topic", bytes)

decodedValue shouldBe MyWrapper("the value") // but was "MyWrapper(value = \"the value\")"
```

**Expected behavior**
Avro STRING should be deserializable to any type.

**Desktop (please complete the following information):**
 - Kotlin version: doesn't apply
 - Avro4k version: 2.4.0

**Additional context**
Idea to fix it: pass a `String.serializer()` for STRING schemas, and then wrap the string to the inline class

However, it may not work well for logical types... to see the best to do

--- a/confluent-kafka-serializer/api/confluent-kafka-serializer.api
+++ b/confluent-kafka-serializer/api/confluent-kafka-serializer.api
@@ -20,8 +20,8 @@ public abstract class com/github/avrokotlin/avro4k/kafka/confluent/AbstractAvro4
 
 public final class com/github/avrokotlin/avro4k/kafka/confluent/GenericAvro4kKafkaDeserializer : com/github/avrokotlin/avro4k/kafka/confluent/AbstractAvro4kKafkaDeserializer {
 	public fun <init> ()V
-	public fun <init> (Lcom/github/avrokotlin/avro4k/Avro;)V
-	public synthetic fun <init> (Lcom/github/avrokotlin/avro4k/Avro;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
+	public fun <init> (Lcom/github/avrokotlin/avro4k/Avro;Lorg/apache/avro/Schema;)V
+	public synthetic fun <init> (Lcom/github/avrokotlin/avro4k/Avro;Lorg/apache/avro/Schema;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
 	public fun <init> (ZLcom/github/avrokotlin/avro4k/Avro;Ljava/util/Map;Lio/confluent/kafka/schemaregistry/client/SchemaRegistryClient;)V
 	public synthetic fun <init> (ZLcom/github/avrokotlin/avro4k/Avro;Ljava/util/Map;Lio/confluent/kafka/schemaregistry/client/SchemaRegistryClient;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
 }
@@ -47,8 +47,8 @@ public final class com/github/avrokotlin/avro4k/kafka/confluent/GenericAvro4kKaf
 
 public final class com/github/avrokotlin/avro4k/kafka/confluent/ReflectAvro4kKafkaDeserializer : com/github/avrokotlin/avro4k/kafka/confluent/AbstractAvro4kKafkaDeserializer {
 	public fun <init> ()V
-	public fun <init> (Lcom/github/avrokotlin/avro4k/Avro;)V
-	public synthetic fun <init> (Lcom/github/avrokotlin/avro4k/Avro;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
+	public fun <init> (Lcom/github/avrokotlin/avro4k/Avro;Lorg/apache/avro/Schema;)V
+	public synthetic fun <init> (Lcom/github/avrokotlin/avro4k/Avro;Lorg/apache/avro/Schema;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
 	public fun <init> (ZLcom/github/avrokotlin/avro4k/Avro;Ljava/util/Map;Lio/confluent/kafka/schemaregistry/client/SchemaRegistryClient;)V
 	public synthetic fun <init> (ZLcom/github/avrokotlin/avro4k/Avro;Ljava/util/Map;Lio/confluent/kafka/schemaregistry/client/SchemaRegistryClient;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
 }
@@ -78,6 +78,7 @@ public final class com/github/avrokotlin/avro4k/kafka/confluent/SpecificAvro4kKa
 	public synthetic fun <init> (Lkotlinx/serialization/DeserializationStrategy;Lcom/github/avrokotlin/avro4k/Avro;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
 	public fun <init> (Lkotlinx/serialization/DeserializationStrategy;ZLcom/github/avrokotlin/avro4k/Avro;Ljava/util/Map;Lio/confluent/kafka/schemaregistry/client/SchemaRegistryClient;)V
 	public synthetic fun <init> (Lkotlinx/serialization/DeserializationStrategy;ZLcom/github/avrokotlin/avro4k/Avro;Ljava/util/Map;Lio/confluent/kafka/schemaregistry/client/SchemaRegistryClient;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
+	public fun deserialize (Ljava/lang/String;Lorg/apache/kafka/common/header/Headers;[B)Ljava/lang/Object;
 }
 
 public class com/github/avrokotlin/avro4k/kafka/confluent/SpecificAvro4kKafkaSerde : org/apache/kafka/common/serialization/Serdes$WrapperSerde {

--- a/confluent-kafka-serializer/src/main/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/AbstractAvro4kKafkaSerde.kt
+++ b/confluent-kafka-serializer/src/main/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/AbstractAvro4kKafkaSerde.kt
@@ -25,6 +25,7 @@ import org.apache.avro.generic.GenericData
 import org.apache.avro.io.DatumReader
 import org.apache.avro.io.DatumWriter
 import org.apache.avro.io.Decoder
+import org.apache.avro.io.Encoder
 import org.apache.avro.io.EncoderFactory
 import org.apache.avro.util.Utf8
 import org.apache.kafka.common.errors.InvalidConfigurationException
@@ -116,16 +117,43 @@ public abstract class AbstractAvro4kKafkaSerializer<T : Any>(
         if (data == null) {
             return null
         }
-        val schema = AvroSchema(getSchema(data))
-        return serializeImpl(getSubjectName(topic, isKey, data, schema), topic, headers, data, schema)
+        val schema = getSchema(data)
+        val data =
+            if (schema.type == Schema.Type.BYTES) {
+                // confluent serializer expects root BYTES to be ByteArray or ByteBuffer, so it is not compatible with value classes wrapping ByteArray or ByteBuffer.
+                // We serialize the data to bytes ourselves here.
+                unwrapRootByteArray(schema, data)
+            } else {
+                data
+            }
+        val avroSchema = AvroSchema(schema)
+        return serializeImpl(getSubjectName(topic, isKey, data, avroSchema), topic, headers, data, avroSchema)
+    }
+
+    private fun unwrapRootByteArray(schema: Schema, data: T): ByteArray {
+        var bytes: ByteArray? = null
+        avro.encodeWithApacheEncoder(
+            schema,
+            serializer,
+            data,
+            object : NotImplementedEncoder() {
+                override fun writeBytes(b: ByteBuffer) {
+                    bytes = b.array()
+                }
+
+                override fun writeBytes(b: ByteArray, start: Int, len: Int) {
+                    bytes = b.copyOfRange(start, start + len)
+                }
+            }
+        )
+        return bytes ?: throw UnsupportedOperationException("avro.encodeWithApacheEncoder did not call writeBytes")
     }
 
     // TODO this will override super method when a new version of the SR is released
     @InternalAvro4kApi
-    // override
     protected open fun getDatumWriter(value: Any, rawSchema: Schema): DatumWriter<Any?> {
         return object : DatumWriter<Any?> {
-            override fun write(value: Any?, out: org.apache.avro.io.Encoder) {
+            override fun write(value: Any?, out: Encoder) {
                 @Suppress("UNCHECKED_CAST")
                 avro.encodeWithApacheEncoder(rawSchema, serializer, value as T, out)
             }
@@ -281,19 +309,17 @@ public abstract class AbstractAvro4kKafkaDeserializer<T : Any>(
     }
 
     override fun getDatumReader(writerSchema: Schema, readerSchema: Schema?): DatumReader<*> {
-        return object : DatumReader<T> {
-            override fun read(reuse: T?, `in`: Decoder): T {
-                return avro.decodeWithApacheDecoder(writerSchema, deserializer, `in`)
-            }
-
-            override fun setSchema(schema: Schema) {
-                throw UnsupportedOperationException()
-            }
-        }
+        // a non-null reader schema means that this.schema returned a schema, so the user wants to always read a specific schema.
+        // confluent takes care of schema adaptations between writer and reader schema, so it always adapt the data to the reader schema
+        // by doing migrations before giving us the data to deserialize.
+        return avro.getDatumReader<Any>(readerSchema ?: writerSchema, deserializer)
     }
 
     @InternalAvro4kApi
     protected abstract val deserializer: DeserializationStrategy<T>
+
+    @InternalAvro4kApi
+    protected abstract val schema: Schema?
 
     override fun deserialize(topic: String?, data: ByteArray?): T? {
         return deserialize(topic, null, data)
@@ -301,10 +327,68 @@ public abstract class AbstractAvro4kKafkaDeserializer<T : Any>(
 
     override fun deserialize(topic: String?, headers: Headers?, data: ByteArray?): T? {
         @Suppress("UNCHECKED_CAST") // getDatumReader is expected to deserialize to T
-        return super<AbstractKafkaAvroDeserializer>.deserialize(topic, isKey, headers, data, null) as T?
+        return super<AbstractKafkaAvroDeserializer>.deserialize(topic, isKey, headers, data, schema) as T?
     }
 
     override fun close() {
         super<AbstractKafkaAvroDeserializer>.close()
     }
+}
+
+internal fun <T> Avro.getDatumReader(readerSchema: Schema, deserializer: DeserializationStrategy<T>): DatumReader<T> {
+    return Avro4kDatumReader(this, readerSchema, deserializer)
+}
+
+internal class Avro4kDatumReader<T>(
+    private val avro: Avro,
+    private val readerSchema: Schema,
+    private val deserializer: DeserializationStrategy<T>,
+) : DatumReader<T> {
+    override fun read(reuse: T?, `in`: Decoder): T {
+        return avro.decodeWithApacheDecoder(readerSchema, deserializer, `in`)
+    }
+
+    override fun setSchema(schema: Schema) {
+        throw UnsupportedOperationException()
+    }
+}
+
+private abstract class NotImplementedEncoder : Encoder() {
+    override fun writeNull(): Unit = throw UnsupportedOperationException()
+
+    override fun writeBoolean(b: Boolean): Unit = throw UnsupportedOperationException()
+
+    override fun writeInt(i: Int): Unit = throw UnsupportedOperationException()
+
+    override fun writeLong(l: Long): Unit = throw UnsupportedOperationException()
+
+    override fun writeFloat(v: Float): Unit = throw UnsupportedOperationException()
+
+    override fun writeDouble(v: Double): Unit = throw UnsupportedOperationException()
+
+    override fun writeString(s: Utf8): Unit = throw UnsupportedOperationException()
+
+    override fun writeBytes(b: ByteBuffer): Unit = throw UnsupportedOperationException()
+
+    override fun writeBytes(b: ByteArray, start: Int, len: Int): Unit = throw UnsupportedOperationException()
+
+    override fun writeFixed(b: ByteArray, start: Int, len: Int): Unit = throw UnsupportedOperationException()
+
+    override fun writeEnum(e: Int): Unit = throw UnsupportedOperationException()
+
+    override fun writeArrayStart(): Unit = throw UnsupportedOperationException()
+
+    override fun setItemCount(itemCount: Long): Unit = throw UnsupportedOperationException()
+
+    override fun startItem(): Unit = throw UnsupportedOperationException()
+
+    override fun writeArrayEnd(): Unit = throw UnsupportedOperationException()
+
+    override fun writeMapStart(): Unit = throw UnsupportedOperationException()
+
+    override fun writeMapEnd(): Unit = throw UnsupportedOperationException()
+
+    override fun writeIndex(i: Int): Unit = throw UnsupportedOperationException()
+
+    override fun flush(): Unit = throw UnsupportedOperationException()
 }

--- a/confluent-kafka-serializer/src/main/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/GenericAvro4kKafkaSerde.kt
+++ b/confluent-kafka-serializer/src/main/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/GenericAvro4kKafkaSerde.kt
@@ -6,6 +6,7 @@ import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient
 import kotlinx.serialization.DeserializationStrategy
 import kotlinx.serialization.SerializationStrategy
 import kotlinx.serialization.serializer
+import org.apache.avro.Schema
 import org.apache.avro.generic.GenericEnumSymbol
 import org.apache.avro.generic.GenericFixed
 import org.apache.avro.generic.GenericRecord
@@ -86,11 +87,14 @@ public class GenericAvro4kKafkaSerializer(avro: Avro = Avro) : AbstractAvro4kKaf
         initialize(schemaRegistry, props, isKey)
     }
 
-    override val serializer: SerializationStrategy<Any> = avro.serializersModule.serializer()
+    override val serializer: SerializationStrategy<Any> = super.avro.serializersModule.serializer()
 }
 
 @ExperimentalAvro4kApi
-public class GenericAvro4kKafkaDeserializer(avro: Avro = Avro) : AbstractAvro4kKafkaDeserializer<Any>(avro.withAnyKSerializer(GenericKSerializer())) {
+public class GenericAvro4kKafkaDeserializer(
+    avro: Avro = Avro,
+    override val schema: Schema? = null,
+) : AbstractAvro4kKafkaDeserializer<Any>(avro.withAnyKSerializer(GenericKSerializer())) {
     /**
      * Create a configured instance of [GenericAvro4kKafkaDeserializer].
      * You need to pass at least a non-null [schemaRegistry], or the `schema.registry.url` entry in [props].
@@ -106,5 +110,5 @@ public class GenericAvro4kKafkaDeserializer(avro: Avro = Avro) : AbstractAvro4kK
         initialize(schemaRegistry, props, isKey)
     }
 
-    override val deserializer: DeserializationStrategy<Any> = avro.serializersModule.serializer()
+    override val deserializer: DeserializationStrategy<Any> = super.avro.serializersModule.serializer()
 }

--- a/confluent-kafka-serializer/src/main/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/ReflectAvro4kKafkaSerde.kt
+++ b/confluent-kafka-serializer/src/main/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/ReflectAvro4kKafkaSerde.kt
@@ -5,6 +5,8 @@ import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
 import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient
 import kotlinx.serialization.DeserializationStrategy
 import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.serializer
+import org.apache.avro.Schema
 import org.apache.kafka.common.serialization.Serdes.WrapperSerde
 
 /**
@@ -80,11 +82,14 @@ public class ReflectAvro4kKafkaSerializer(avro: Avro = Avro) : AbstractAvro4kKaf
         initialize(schemaRegistry, props, isKey)
     }
 
-    override val serializer: SerializationStrategy<Any> = ReflectKSerializer()
+    override val serializer: SerializationStrategy<Any> = super.avro.serializersModule.serializer()
 }
 
 @ExperimentalAvro4kApi
-public class ReflectAvro4kKafkaDeserializer(avro: Avro = Avro) : AbstractAvro4kKafkaDeserializer<Any>(avro.withAnyKSerializer(ReflectKSerializer())) {
+public class ReflectAvro4kKafkaDeserializer(
+    avro: Avro = Avro,
+    override val schema: Schema? = null,
+) : AbstractAvro4kKafkaDeserializer<Any>(avro.withAnyKSerializer(ReflectKSerializer())) {
     /**
      * Create a configured instance of [ReflectAvro4kKafkaDeserializer].
      * You need to pass at least a non-null [schemaRegistry], or the `schema.registry.url` entry in [props].
@@ -98,5 +103,5 @@ public class ReflectAvro4kKafkaDeserializer(avro: Avro = Avro) : AbstractAvro4kK
         initialize(schemaRegistry, props, isKey)
     }
 
-    override val deserializer: DeserializationStrategy<Any> = ReflectKSerializer()
+    override val deserializer: DeserializationStrategy<Any> = super.avro.serializersModule.serializer()
 }

--- a/confluent-kafka-serializer/src/main/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/SpecificAvro4kKafkaSerde.kt
+++ b/confluent-kafka-serializer/src/main/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/SpecificAvro4kKafkaSerde.kt
@@ -3,18 +3,26 @@ package com.github.avrokotlin.avro4k.kafka.confluent
 import com.github.avrokotlin.avro4k.Avro
 import com.github.avrokotlin.avro4k.ExperimentalAvro4kApi
 import com.github.avrokotlin.avro4k.internal.aliases
+import com.github.avrokotlin.avro4k.internal.decodeWithApacheDecoder
 import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient
 import io.confluent.kafka.serializers.KafkaAvroDeserializerConfig
 import kotlinx.serialization.DeserializationStrategy
 import kotlinx.serialization.ExperimentalSerializationApi
 import kotlinx.serialization.KSerializer
 import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.builtins.ByteArraySerializer
+import kotlinx.serialization.builtins.serializer
 import kotlinx.serialization.descriptors.nonNullOriginal
 import kotlinx.serialization.serializer
 import org.apache.avro.Schema
+import org.apache.avro.io.DatumReader
+import org.apache.avro.io.Decoder
+import org.apache.avro.util.Utf8
 import org.apache.kafka.common.config.ConfigException
+import org.apache.kafka.common.header.Headers
 import org.apache.kafka.common.serialization.Serde
 import org.apache.kafka.common.serialization.Serdes.WrapperSerde
+import java.nio.ByteBuffer
 import kotlin.reflect.KClass
 
 /**
@@ -151,10 +159,12 @@ public class SpecificAvro4kKafkaDeserializer<T : Any>(
     avro: Avro = Avro,
 ) : AbstractAvro4kKafkaDeserializer<T>(avro.withAnyKSerializer(ReflectKSerializer())) {
     override lateinit var deserializer: DeserializationStrategy<T> private set
+    override lateinit var schema: Schema private set
+    private val currentWriterSchema = ThreadLocal<Schema>()
 
     init {
         if (deserializer != null) {
-            this.deserializer = deserializer
+            setDeserializer(deserializer)
         }
     }
 
@@ -192,10 +202,16 @@ public class SpecificAvro4kKafkaDeserializer<T : Any>(
                         "The configured type name was not found "
                     )
             @Suppress("UNCHECKED_CAST")
-            deserializer = foundKSerializer as DeserializationStrategy<T>
+            setDeserializer(foundKSerializer as DeserializationStrategy<T>)
         }
 
         super.configure(config)
+    }
+
+    private fun setDeserializer(deserializer: DeserializationStrategy<T>) {
+        this.deserializer = deserializer
+        @OptIn(ExperimentalSerializationApi::class)
+        schema = avro.schema(deserializer.descriptor.nonNullOriginal)
     }
 
     @OptIn(ExperimentalSerializationApi::class)
@@ -208,6 +224,61 @@ public class SpecificAvro4kKafkaDeserializer<T : Any>(
             avro.serializersModule.dumpTo(it)
             it.found
         }
+    }
+
+    override fun getDatumReader(writerSchema: Schema, readerSchema: Schema?): DatumReader<*> {
+        when (writerSchema.type) {
+            Schema.Type.STRING -> {
+                currentWriterSchema.set(writerSchema)
+                // confluent applies result.toString() for root STRING schemas
+                // so we force the DatumReader to return a String, and then we will decode it properly in deserialize() using the given k-deserializer
+                return avro.getDatumReader(writerSchema, String.serializer())
+            }
+
+            Schema.Type.BYTES -> {
+                currentWriterSchema.set(writerSchema)
+                // confluent bypasses the DatumReader, always returning ByteArray for root BYTES schemas
+                // so we force the DatumReader to return a ByteArray, and then we will decode it properly in deserialize() using the given k-deserializer
+                return avro.getDatumReader(writerSchema, ByteArraySerializer())
+            }
+
+            else -> return super.getDatumReader(writerSchema, readerSchema)
+        }
+    }
+
+    override fun deserialize(topic: String?, headers: Headers?, data: ByteArray?): T? {
+        val result = super.deserialize(topic, headers, data)
+        currentWriterSchema.get()?.let { writerSchema ->
+            currentWriterSchema.set(null)
+            return when (writerSchema.type) {
+                Schema.Type.STRING -> wrapRootString(result as String, writerSchema)
+                Schema.Type.BYTES -> wrapRootByteArray(result as ByteArray, writerSchema)
+                else -> result
+            }
+        }
+        return result
+    }
+
+    private fun wrapRootString(rootString: String, writerSchema: Schema): T {
+        return avro.decodeWithApacheDecoder(
+            writerSchema,
+            deserializer,
+            object : NoImplementedDecoder() {
+                override fun readString(old: Utf8?) = Utf8(rootString)
+
+                override fun readString() = rootString
+            }
+        )
+    }
+
+    private fun wrapRootByteArray(rootBytes: ByteArray, writerSchema: Schema): T {
+        return avro.decodeWithApacheDecoder(
+            writerSchema,
+            deserializer,
+            object : NoImplementedDecoder() {
+                override fun readBytes(old: ByteBuffer?) = ByteBuffer.wrap(rootBytes)
+            }
+        )
     }
 }
 
@@ -237,4 +308,48 @@ private class KSerializerFinder(
     override fun <Base : Any> polymorphicDefaultDeserializer(baseClass: KClass<Base>, defaultDeserializerProvider: (className: String?) -> DeserializationStrategy<Base>?) {
         defaultDeserializerProvider(serialName)?.let { setFoundKSerializer(it) }
     }
+}
+
+private abstract class NoImplementedDecoder : Decoder() {
+    override fun readNull() = throw UnsupportedOperationException()
+
+    override fun readBoolean(): Boolean = throw UnsupportedOperationException()
+
+    override fun readInt(): Int = throw UnsupportedOperationException()
+
+    override fun readLong(): Long = throw UnsupportedOperationException()
+
+    override fun readFloat(): Float = throw UnsupportedOperationException()
+
+    override fun readDouble(): Double = throw UnsupportedOperationException()
+
+    override fun readString(old: Utf8?): Utf8? = throw UnsupportedOperationException()
+
+    override fun readString(): String? = throw UnsupportedOperationException()
+
+    override fun skipString() = throw UnsupportedOperationException()
+
+    override fun readBytes(old: ByteBuffer?): ByteBuffer? = throw UnsupportedOperationException()
+
+    override fun skipBytes() = throw UnsupportedOperationException()
+
+    override fun readFixed(bytes: ByteArray, start: Int, length: Int) = throw UnsupportedOperationException()
+
+    override fun skipFixed(length: Int) = throw UnsupportedOperationException()
+
+    override fun readEnum(): Int = throw UnsupportedOperationException()
+
+    override fun readArrayStart(): Long = throw UnsupportedOperationException()
+
+    override fun arrayNext(): Long = throw UnsupportedOperationException()
+
+    override fun skipArray(): Long = throw UnsupportedOperationException()
+
+    override fun readMapStart(): Long = throw UnsupportedOperationException()
+
+    override fun mapNext(): Long = throw UnsupportedOperationException()
+
+    override fun skipMap(): Long = throw UnsupportedOperationException()
+
+    override fun readIndex(): Int = throw UnsupportedOperationException()
 }

--- a/confluent-kafka-serializer/src/test/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/AbstractAvro4kKafkaDeserializerTest.kt
+++ b/confluent-kafka-serializer/src/test/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/AbstractAvro4kKafkaDeserializerTest.kt
@@ -1,6 +1,7 @@
 package com.github.avrokotlin.avro4k.kafka.confluent
 
 import com.github.avrokotlin.avro4k.Avro
+import com.github.avrokotlin.avro4k.InternalAvro4kApi
 import io.confluent.kafka.schemaregistry.avro.AvroSchema
 import io.confluent.kafka.schemaregistry.client.MockSchemaRegistryClient
 import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient
@@ -36,6 +37,10 @@ private class MockedDeserializer(
     props: Map<String, *> = emptyMap<String, Any>(),
 ) : AbstractAvro4kKafkaDeserializer<Any>(avro) {
     override val deserializer: DeserializationStrategy<Any> = mockk<DeserializationStrategy<Any>>()
+
+    @InternalAvro4kApi
+    override val schema: Schema?
+        get() = null
 
     init {
         every { deserializer.deserialize(any()) } returns decodedValue

--- a/confluent-kafka-serializer/src/test/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/AbstractAvro4kKafkaSerializerTest.kt
+++ b/confluent-kafka-serializer/src/test/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/AbstractAvro4kKafkaSerializerTest.kt
@@ -53,7 +53,7 @@ class AbstractAvro4kKafkaSerializerTest : StringSpec() {
             TestData("test", 24) to Avro.schema<TestData>(),
             SomeEnum.B to Avro.schema<SomeEnum>(),
             listOf("one", "two", 3) to Schema.createArray(Schema.create(Schema.Type.STRING)),
-            setOf(1, 2.0, 3f) to Schema.createArray(Schema.create(Schema.Type.INT)),
+            setOf<Any>(1, 2L, 3L) to Schema.createArray(Schema.create(Schema.Type.INT)),
             arrayOf(1, 2, 3) to Schema.createArray(Schema.create(Schema.Type.INT)),
             Utf8("string") to Schema.create(Schema.Type.STRING),
             "string" to Schema.create(Schema.Type.STRING),
@@ -84,13 +84,14 @@ class AbstractAvro4kKafkaSerializerTest : StringSpec() {
     private infix fun Any.shouldRegisterSchema(
         expectedSchema: Schema,
     ) {
-        val serializer = MockedSerializer()
+        val schemaRegistry = MockSchemaRegistryClient()
+        val serializer = GenericAvro4kKafkaSerializer(isKey = true, schemaRegistry = schemaRegistry)
 
         serializer.serialize("topic", this)
 
-        serializer.schemaRegistry.allSubjects shouldBe listOf("topic-value")
-        serializer.schemaRegistry.getAllVersions("topic-value") shouldBe listOf(1)
-        serializer.schemaRegistry.getSchemaById(1).rawSchema() shouldBe expectedSchema
+        schemaRegistry.allSubjects shouldBe listOf("topic-key")
+        schemaRegistry.getAllVersions("topic-key") shouldBe listOf(1)
+        schemaRegistry.getSchemaById(1).rawSchema() shouldBe expectedSchema
     }
 }
 

--- a/confluent-kafka-serializer/src/test/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/SpecificAvro4kKafkaDeserializerTest.kt
+++ b/confluent-kafka-serializer/src/test/kotlin/com/github/avrokotlin/avro4k/kafka/confluent/SpecificAvro4kKafkaDeserializerTest.kt
@@ -1,0 +1,67 @@
+package com.github.avrokotlin.avro4k.kafka.confluent
+
+import io.confluent.kafka.schemaregistry.client.MockSchemaRegistryClient
+import io.kotest.core.spec.style.StringSpec
+import io.kotest.matchers.should
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.types.beInstanceOf
+import kotlinx.serialization.Serializable
+
+class SpecificAvro4kKafkaDeserializerTest : StringSpec() {
+    init {
+        "value class of String should be deserializable" {
+            val schemaRegistry = MockSchemaRegistryClient()
+            val bytes = GenericAvro4kKafkaSerializer(isKey = true, schemaRegistry = schemaRegistry).serialize("topic", "test data")
+            val deserializer = SpecificAvro4kKafkaDeserializer(StringWrapper.serializer(), isKey = true, schemaRegistry = schemaRegistry)
+
+            val deserialized = deserializer.deserialize("topic", bytes)
+
+            deserialized shouldBe StringWrapper("test data")
+        }
+
+        "value class of ByteArray should be serializable" {
+            val schemaRegistry = MockSchemaRegistryClient()
+            val serializer = SpecificAvro4kKafkaSerializer<BytesWrapper>(isKey = true, schemaRegistry = schemaRegistry)
+
+            val bytes = serializer.serialize("topic", BytesWrapper(byteArrayOf(1, 2, 3)))
+
+            val deserialized = GenericAvro4kKafkaDeserializer(isKey = true, schemaRegistry = schemaRegistry).deserialize("topic", bytes)
+            deserialized shouldBe byteArrayOf(1, 2, 3)
+        }
+
+        "value class of ByteArray should be deserializable" {
+            val schemaRegistry = MockSchemaRegistryClient()
+            val bytes = GenericAvro4kKafkaSerializer(isKey = true, schemaRegistry = schemaRegistry).serialize("topic", byteArrayOf(1, 2, 3))
+            val deserializer = SpecificAvro4kKafkaDeserializer(BytesWrapper.serializer(), isKey = true, schemaRegistry = schemaRegistry)
+
+            val deserialized = deserializer.deserialize("topic", bytes)
+
+            deserialized should beInstanceOf<BytesWrapper>()
+            deserialized!!.value shouldBe byteArrayOf(1, 2, 3)
+        }
+
+        "enum should be deserializable from a serialized string" {
+            val schemaRegistry = MockSchemaRegistryClient()
+            val bytes = GenericAvro4kKafkaSerializer(isKey = true, schemaRegistry = schemaRegistry).serialize("topic", "B")
+            val deserializer = SpecificAvro4kKafkaDeserializer(AnEnum.serializer(), isKey = true, schemaRegistry = schemaRegistry)
+
+            val deserialized = deserializer.deserialize("topic", bytes)
+
+            deserialized shouldBe AnEnum.B
+        }
+    }
+}
+
+@JvmInline
+@Serializable
+private value class StringWrapper(val value: String)
+
+@JvmInline
+@Serializable
+private value class BytesWrapper(val value: ByteArray)
+
+@Serializable
+private enum class AnEnum {
+    A,
+    B,
+}

